scheme/parsec-tpm/evidence_handler.go

```diff
@@ -104,3 +225,185 @@ func parsecTpmLookupKey(scope, tenantID, class, instance string) string {
 
 	return u.String()
 }
+
+func evidenceToMap(e parsectpm.Evidence) (map[string]interface{}, error) {
+	data, err := e.ToJSON()
+	if err != nil {
+		return nil, err
+	}
+
+	var out map[string]interface{}
+	err = json.Unmarshal(data, &out)
+
+	return out, err
+}
+
+func populateAttestationResult(
+	result *ear.AttestationResult,
+	evidence map[string]interface{},
+	endorsements []Endorsements,
+) error {
+	appraisal := result.Submods[SchemeName]
+
+	// once the signature on the token is verified, we can claim the HW is
+	// authentic
+	appraisal.TrustVector.Hardware = ear.GenuineHardwareClaim
+
+	ev, err := mapToEvidence(evidence)
+	if err != nil {
+		return fmt.Errorf("failed attestation results: %w", err)
+	}
+	if ev.Pat == nil {
+		return errors.New("no platform token to evaluate")
+	}
+
+	attInfo, err := ev.Pat.GetAttestationInfo()
+	if err != nil {
+		return fmt.Errorf("unable to get attestation information: %w", err)
+	}
+	if attInfo == nil {
+		return errors.New("attestation information in platform token is nil")
+	}
+
+	pcrs := attInfo.PCR.PCRinfo.PCRs
+	hashAlgID := attInfo.PCR.PCRinfo.HashAlgID
+	pcrDigest := attInfo.PCR.PCRDigest
+
+	// Match the Evidence PCR against the endorsements to generate a subset
+	// of matching endorsements
+	eds, err := matchPCRs(pcrs, hashAlgID, endorsements)
+	if err != nil {
+		appraisal.TrustVector.Executables = ear.UnrecognizedRuntimeClaim
+		log.Error("match PCR failed: %w", err)
+		return fmt.Errorf("match PCR failed: %w", err)
+	}
+
+	if err := matchPCRDigest(pcrDigest, hashAlgID, eds); err != nil {
+		appraisal.TrustVector.Executables = ear.UnrecognizedRuntimeClaim
+		log.Error("match PCR Digest failed: %w", err)
+		return fmt.Errorf("match failed for PCR Digest: %w", err)
+	}
+
+	appraisal.TrustVector.Executables = ear.ApprovedRuntimeClaim
+	appraisal.TrustVector.InstanceIdentity = ear.TrustworthyInstanceClaim
+	log.Debug("matchPCRs and matchPCRDigest Success")
+
+	// Populate Veraison Key Attestation Extension
+	if ev.Kat == nil {
```

Comment:
if KAT's presence is non-optional, it should be checked earlier to avoid going through the fairly costly match algorithm that we would trash anyway.